src/MagnumPlugins/DdsImporter/DdsImporter.cpp

```diff
@@ -0,0 +1,387 @@
+        for (UnsignedInt i = 0; i < numMipmaps && (w || h); i++) {
+            /* shrink to next power of 2 */
+            w = Math::max(w >> 1, 1u);
+            h = Math::max(h >> 1, 1u);
+            d = Math::max(d >> 1, 1u);
+
+            if (compressed) {
+                loadCompressedImageData(colorFormat.compressed, w, h, d, dimensions);
+            } else {
+                loadUncompressedImageData(colorFormat.uncompressed, w, h, d, components, dimensions);
+            }
+        }
+    }
+
+}
+
+void DdsImporter::loadUncompressedImageData(const ColorFormat format,
+    const UnsignedInt width, const UnsignedInt height, const UnsignedInt depth,
+    const UnsignedInt components, const UnsignedByte dimensions)
+{
+    const unsigned int numPixels = width * height * depth;
+    const unsigned int size = numPixels * components;
+
+    /* load image data */
+    char *data = new char[size];
+    _in->read(data, size);
+
+    ColorFormat newColorFormat = format;
+    if (format == ColorFormat::BGR) {
+        Debug() << "Converting from BGR to RGB";
+        convertBGRToRGB(data, numPixels);
+        newColorFormat = ColorFormat::RGB;
+    } else if (format == ColorFormat::BGRA) {
+        Debug() << "Converting from BGRA to RGBA";
+        convertBGRAToRGBA(data, numPixels);
+        newColorFormat = ColorFormat::RGBA;
+    }
+
+    if (dimensions == 2) {
+        _imageData2D.emplace_back(newColorFormat, ColorType::UnsignedByte, Vector2i{Int(width), Int(height)}, std::move(static_cast<void*>(data)));
+    } else if (dimensions == 3) {
+        _imageData3D.emplace_back(newColorFormat, ColorType::UnsignedByte, Vector3i{Int(width), Int(height), Int(depth)}, std::move(static_cast<void*>(data)));
+    } else {
+        Error() << "Unsupported image dimensions:" << dimensions;
+    }
+}
+
+void DdsImporter::loadCompressedImageData(const CompressedColorFormat format,
+    const UnsignedInt width, const UnsignedInt height,
+    const UnsignedInt depth, const UnsignedByte dimensions)
+{
+    const unsigned int size = (depth*((width + 3)/4)*(((height + 3)/4))*((format == CompressedColorFormat::RGBAS3tcDxt1) ? 8 : 16));
+
+    /* load image data */
+    char *data = new char[size];
+    _in->read(data, size);
+
+    if (dimensions == 2) {
+        _imageData2D.emplace_back(format, Vector2i{Int(width), Int(height)}, Containers::Array<char>(data, size));
+    } else if (dimensions == 3) {
+        _imageData3D.emplace_back(format, Vector3i{Int(width), Int(height), Int(depth)}, Containers::Array<char>(data, size));
+    } else {
+        Error() << "Unsupported image dimensions:" << dimensions;
+    }
+}
+
+void DdsImporter::doOpenFile(const std::string& filename) {
+    /* let doOpenData handle it */
+    AbstractImporter::doOpenFile(filename);
+}
+
+UnsignedInt DdsImporter::doImage2DCount() const { return _imageData2D.size(); }
+
+std::optional<ImageData2D> DdsImporter::doImage2D(UnsignedInt id) {
+    if (id < _imageData2D.size()) {
```

Comment:
This check is not needed (the `AbstractImporter` does this already using information from `doImage2DCount()`), similarly for the 3D image below.